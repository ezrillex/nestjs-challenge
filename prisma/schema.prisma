// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// Looking for ways to speed up your queries, or scale easily with your serverless or edge functions?
// Try Prisma Accelerate: https://pris.ly/cli/accelerate-init

// npx prisma migrate dev --name the_name
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Testing {
  id   String @id @default(uuid())
  test String
}

enum roles {
  customer
  manager
  admin
  anonymous // public
}

model Users {
  id                                 String     @id @default(uuid())
  first_name                         String     @db.VarChar(60)
  last_name                          String     @db.VarChar(120)
  email                              String     @unique
  password                           String     @db.Char(60) // hashed and salted password
  created_at                         DateTime   @default(now())
  role                               roles      @default(customer)
  failed_login_attempts              Int        @default(0)
  failed_login_attempts_timestamps   DateTime[] @default([])
  last_login_at                      DateTime?
  password_reset_requests            Int        @default(0)
  password_reset_requests_timestamps DateTime[] @default([])
  password_reset_token               String?
  password_last_updated              DateTime   @default(now())
}
