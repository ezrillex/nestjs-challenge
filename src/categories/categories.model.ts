import { Field, ObjectType } from '@nestjs/graphql';
import { Products } from '../products/products.model';

@ObjectType()
export class Categories {
  @Field(() => String)
  id: string;

  @Field(() => String)
  name: string;
}

@ObjectType()
export class GetCategoriesResponse extends Categories {
  @Field(() => [Products])
  products?: Products[];
}
