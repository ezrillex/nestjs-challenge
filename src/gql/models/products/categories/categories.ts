import { Field, ObjectType } from '@nestjs/graphql';
import { Products } from '../products.model';

@ObjectType()
export class Categories {
  @Field(() => String)
  id: string;

  @Field(() => String)
  name: string;

  @Field(() => [Products])
  products: Products[];
}
