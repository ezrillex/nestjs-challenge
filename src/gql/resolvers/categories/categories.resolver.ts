import { Args, Mutation, Resolver } from '@nestjs/graphql';
import { RequiresRole } from '../../../decorators/requires-role/requires-role.decorator';
import { roles } from '@prisma/client';
import { CreateCategoryInput } from '../../models/products/create-category-input/create-category-input';
import { ProductsService } from '../../../services/products/products.service';

@Resolver()
export class CategoriesResolver {
  // todo refactor this to be its own service? its too small tbh not want to do it as it bloats the app?
  constructor(private productsService: ProductsService) {}

  @RequiresRole(roles.manager)
  @Mutation(() => String, { nullable: true })
  async createCategory(
    @Args('CreateCategoryInput') createCategoryInput: CreateCategoryInput,
  ) {
    const result =
      await this.productsService.CreateCategory(createCategoryInput);
    return result.id;
  }
}
