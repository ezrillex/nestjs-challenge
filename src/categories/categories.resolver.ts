import {
  Args,
  Context,
  ID,
  Mutation,
  Parent,
  Query,
  ResolveField,
  Resolver,
} from '@nestjs/graphql';
import { RequiresRole } from '../common/decorators/requires-role.decorator';
import { roles } from '@prisma/client';
import { ParseUUIDPipe } from '@nestjs/common';
import { CategoriesService } from './categories.service';
import { Categories, GetCategoriesResponse } from './categories.model';
import { PublicPrivate } from '../common/decorators/public_and_private.decorator';
import { Products } from '../products/products.model';

@Resolver(() => GetCategoriesResponse)
export class CategoriesResolver {
  constructor(private categoriesService: CategoriesService) {}

  @RequiresRole(roles.manager)
  @Mutation(() => Categories, { nullable: true })
  async createCategory(@Args('name') name: string): Promise<Categories> {
    return this.categoriesService.createCategory(name);
  }

  @PublicPrivate()
  @Query(() => [GetCategoriesResponse], { nullable: true })
  async getCategories(): Promise<Categories[]> {
    return this.categoriesService.getAllCategories();
  }

  @ResolveField()
  async products(
    @Parent() category: Categories,
    @Context('req') request: Request,
  ): Promise<Products[]> {
    const { id } = category;
    return this.categoriesService.getProductsByCategory(
      id,
      request['user'].role,
    );
  }

  @RequiresRole(roles.manager)
  @Mutation(() => String, { nullable: true })
  async deleteCategory(
    @Args('id', { type: () => ID }, ParseUUIDPipe) id: string,
  ): Promise<string> {
    return this.categoriesService.deleteCategoryById(id);
  }
}
