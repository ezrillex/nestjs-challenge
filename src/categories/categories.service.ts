import {
  Injectable,
  NotAcceptableException,
  NotFoundException,
} from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { Categories } from './categories.model';
import { Products } from '../products/products.model';
import { roles } from '@prisma/client';

@Injectable()
export class CategoriesService {
  constructor(private readonly prisma: PrismaService) {}

  async createCategory(data: string): Promise<Categories> {
    const exists = await this.prisma.categories.count({
      where: { name: data },
    });
    if (exists > 0) {
      throw new NotAcceptableException('Category already exists');
    }

    return this.prisma.categories.create({
      data: {
        name: data,
      },
    });
  }

  async getProductsByCategory(id: string, role: roles): Promise<Products[]> {
    let hide = {};
    if (role === roles.manager) {
      hide = {
        where: {
          is_deleted: false,
        },
      };
    } else {
      hide = {
        where: {
          is_deleted: false,
          is_published: true,
        },
      };
    }

    const { Products } = await this.prisma.categories.findUnique({
      where: {
        id,
      },
      select: {
        Products: hide,
      },
    });
    return Products;
  }

  async getAllCategories(): Promise<Categories[]> {
    return this.prisma.categories.findMany({});
  }

  async getCategoriesByProduct(id: string): Promise<Categories[]> {
    const { categories } = await this.prisma.products.findUnique({
      where: {
        id: id,
      },
      select: {
        categories: true,
      },
    });
    return categories;
  }

  async deleteCategoryById(id: string): Promise<string> {
    const record = await this.prisma.categories.count({
      where: { id: id },
    });

    if (record === 0) {
      throw new NotFoundException('Record not found.');
    }

    await this.prisma.categories.delete({
      where: {
        id: id,
      },
    });
    return 'Record Deleted';
  }
}
