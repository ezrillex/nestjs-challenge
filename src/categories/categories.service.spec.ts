import { Test, TestingModule } from '@nestjs/testing';
import { CategoriesService } from './categories.service';
import { PrismaService } from '../prisma/prisma.service';
import { Categories } from '@prisma/client';

describe('CategoriesService', () => {
  let service: CategoriesService;
  let prismaService: PrismaService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [PrismaService, CategoriesService],
    }).compile();

    service = module.get<CategoriesService>(CategoriesService);
    prismaService = module.get<PrismaService>(PrismaService);
  });

  it('CategoriesService should be defined', async () => {
    expect(service).toBeDefined();
  });

  it('PrismaService should be defined', async () => {
    expect(prismaService).toBeDefined();
  });

  it('createCategory Should be defined', async () => {
    expect(service.createCategory).toBeDefined();
  });

  it('getProductsByCategory Should be defined', async () => {
    expect(service.getProductsByCategory).toBeDefined();
  });

  it('getAllCategories Should be defined', async () => {
    expect(service.getAllCategories).toBeDefined();
  });

  it('getCategoriesByProduct Should be defined', async () => {
    expect(service.getCategoriesByProduct).toBeDefined();
  });

  it('deleteCategoryById should be defined', async () => {
    expect(service.deleteCategoryById).toBeDefined();
  });

  describe('createCategory', () => {
    it('should throw if category name already exists', async () => {
      jest.spyOn(prismaService.categories, 'count').mockResolvedValue(1);

      await expect(
        service.createCategory('some name that definitely is not duped'),
      ).rejects.toThrowErrorMatchingSnapshot('category already exists error');
    });

    it('should create a category ', async () => {
      let created = false;
      jest.spyOn(prismaService.categories, 'count').mockResolvedValue(0);
      jest.spyOn(prismaService.categories, 'create').mockImplementation(() => {
        created = true;
        return null;
      });
      await service.createCategory('some name that definitely is not duped');
      expect(created).toEqual(true);
    });
  });

  describe('getProductsByCategory', () => {});

  describe('getAllCategories', () => {});

  describe('getCategoriesByProduct', () => {});

  describe('deleteCategoryById', () => {
    it('should throw if category name already exists', async () => {
      await expect(
        service.deleteCategoryById('2730fc05-6f87-49e5-8a41-559208048ebe'),
      ).rejects.toThrowErrorMatchingSnapshot('category does not exist');
    });

    it('should return a specific message when successful ', async () => {
      jest.spyOn(prismaService.categories, 'count').mockResolvedValue(1);
      jest.spyOn(prismaService.categories, 'delete').mockResolvedValue({
        id: '2730fc05-6f87-49e5-8a41-559208048ebe',
      } as Categories);
      await expect(
        service.deleteCategoryById('2730fc05-6f87-49e5-8a41-559208048ebe'),
      ).resolves.toEqual('Record Deleted');
    });

    it('should throw if prisma fails', async () => {
      jest.spyOn(prismaService.categories, 'count').mockResolvedValue(1);
      jest
        .spyOn(prismaService.categories, 'delete')
        .mockRejectedValue(new Error('testing'));
      await expect(
        service.deleteCategoryById('2730fc05-6f87-49e5-8a41-559208048ebe'),
      ).rejects.toThrowErrorMatchingSnapshot('error when deleting');
    });
  });
});
