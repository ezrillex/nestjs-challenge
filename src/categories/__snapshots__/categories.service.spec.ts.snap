// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesService createCategory should throw if category name already exists: category already exists error 1`] = `"Category already exists"`;

exports[`CategoriesService deleteCategoryById should throw if category name already exists: category does not exist 1`] = `"Record not found."`;

exports[`CategoriesService deleteCategoryById should throw if prisma fails: error when deleting 1`] = `"testing"`;
